rpc/baremetalhostagent/agent/moc_baremetalhost_agent.proto

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache v2.0 license.
+
+syntax = "proto3";
+option go_package = "github.com/microsoft/moc/rpc/nodeagent/compute";
+
+package moc.nodeagent.compute;
+import "google/protobuf/wrappers.proto";
+import "moc_common_common.proto";
+import "moc_common_computecommon.proto";
+
+
+message BareMetalHostSSHPublicKey {
+	string keydata = 1;
+}
+
+message BareMetalHostDisk {
+	string diskName = 1;
+	int32 diskSizeGB = 2;
+}
+
+message BareMetalHostStorageConfiguration {
+	repeated BareMetalHostDisk disks = 1;
+	string imageReference = 2;
+}
+
+message BareMetalHostNetworkInterface {
+	// reference to network interface managed by network
+	string networkInterfaceName = 1;
+}
+
+message BareMetalHostNetworkConfiguration {
+	repeated BareMetalHostNetworkInterface interfaces = 1;
+}
+
+message BareMetalHostUserConfiguration {
+	string username = 1;
+	string password = 2;
+	UserType usertype = 3;
+}
+
+message BareMetalHostRDPConfiguration {
+	bool disableRDP = 1;
+}
+
+message BareMetalHostLinuxConfiguration {
+	bool disablePasswordAuthentication = 1;
+}
+
+message BareMetalHostWindowsConfiguration {
+	bool enableAutomaticUpdates = 1;
+	string timeZone = 2;
+	BareMetalHostRDPConfiguration RDPConfiguration = 3;
+}
+
+message BareMetalHostOperatingSystemConfiguration {
+	string computerName = 1;
+	BareMetalHostUserConfiguration administrator = 2;
+	repeated BareMetalHostUserConfiguration users = 3;
+	string customData = 4;
+	repeated BareMetalHostSSHPublicKey publickeys = 5;
+	OperatingSystemType ostype = 6;
+	// bootstrap engine can be cloud-init, Windows answer files, ...
+	OperatingSystemBootstrapEngine osBootstrapEngine = 7;
+	BareMetalHostLinuxConfiguration linuxConfiguration = 8;
+	BareMetalHostWindowsConfiguration windowsConfiguration = 9;
+}
+
+message BareMetalMachineSize {
+	int32 cpuCount = 1;
+	int32 gpuCount = 2;
+	int32 memoryMB = 3;
+}
+
+message BareMetalHostHardwareConfiguration {
+	BareMetalMachineSize size = 1;
+}
+
+message BareMetalHostSecurityConfiguration {
+	bool enableTPM = 1;
+}
+
+message BareMetalHost {
```

Comment:
One of the things I really hate about the ECP code design, is the fact that some of these fields are user provided properties and some of them are reported status values. It would be kind of nice if we could cleanly separate them. Though this would create inconsistency in the ECP codebase, which would also be bad.

Reply:
Wonder if a way out of this is to have 2 entities. One tracking the desired state on the machine, the other tracking the machine itself. I recall Madhan mentioned the second entity in one of the earlier conversations as the OsDeployment (my name, but same concept).

Reply:
FYI: While bare-metal machines should have two entities in cloud-agent (one for the machine registration and one for the OS deployment), my comment here is a slightly different issue.

This comment is saying that for a given entity in the gRPC API, the settings and the state fields should live in different `message`s. In theory, this will make the API easier to understand.

Reply:
Sure. Are the messages usually not mirroring the entities though? So if we have separate entities, we would get separate messages (if we follow the usual pattern for messages).